mockrxandroidble/src/main/java/com/polidea/rxandroidble2/mockrxandroidble/RxBleClientMock.java

```diff
@@ -363,7 +389,126 @@ private static boolean filterDevice(RxBleDevice rxBleDevice, @Nullable UUID[] fi
+            return true;
+        }
+        RxBleDeviceMock mock = (RxBleDeviceMock)rxBleDevice;
+        ScanRecord scanRecord = mock.getScanRecord();
+        if(scanRecord == null) {
+            return false;
+        }
+
+        String mac = mock.getMacAddress();
+        String name = scanRecord.getDeviceName();
+        Set<UUID> advertisedUUIDs = new HashSet<>(mock.getAdvertisedUUIDs());
+        List<ParcelUuid> scanRecordAdvertisedUUIDs = scanRecord.getServiceUuids();
+        if(scanRecordAdvertisedUUIDs != null) {
+            for (ParcelUuid uuid : scanRecordAdvertisedUUIDs) {
+                advertisedUUIDs.add(uuid.getUuid());
+            }
+        }
+
+        for (ScanFilter filter : scanFilters) {
+            ParcelUuid serviceUUIDMask = filter.getServiceUuidMask();
+            ParcelUuid serviceUUID = filter.getServiceUuid();
+            if(serviceUUIDMask != null && serviceUUID != null) {
+                byte[] serviceUUIDMaskData = RxBleClientMock.getDataFromUUID(serviceUUIDMask.getUuid());
+                byte[] serviceUUIDData = RxBleClientMock.getDataFromUUID(serviceUUID.getUuid());
+                boolean found = false;
+                for (UUID uuid: advertisedUUIDs) {
+                    byte[] UUIDData = RxBleClientMock.getDataFromUUID(uuid);
+                    if(RxBleClientMock.maskedDataEquals(serviceUUIDData, UUIDData, serviceUUIDMaskData)) {
+                        found = true;
+                        break;
+                    }
+                }
+                if(!found) {
+                    return false;
+                }
+            } else if(serviceUUID != null && !advertisedUUIDs.contains(serviceUUID.getUuid())) {
+                return false;
+            }
+
+            if(filter.getDeviceAddress() != null && !filter.getDeviceAddress().equals(mac)) {
+                return false;
+            }
+
+            if(filter.getDeviceName() != null && !filter.getDeviceName().equals(name)) {
+                return false;
+            }
+
+            byte[] manuFilterData = filter.getManufacturerData();
+            if(manuFilterData != null) {
+                byte[] manuData = scanRecord.getManufacturerSpecificData(filter.getManufacturerId());
+                if(manuData == null || !RxBleClientMock.maskedDataEquals(manuData, manuFilterData, filter.getManufacturerDataMask())) {
+                    return false;
+                }
+            }
+            byte[] filterServiceData = filter.getServiceData();
+            ParcelUuid serviceDataUuid = filter.getServiceDataUuid();
+            if(serviceDataUuid != null && filterServiceData != null) {
+                byte[] serviceData = scanRecord.getServiceData(serviceDataUuid);
+                if(serviceData == null || !RxBleClientMock.maskedDataEquals(filterServiceData, serviceData, filter.getServiceDataMask())) {
+                    return false;
+                }
+            }
+        }
+
+        return true;
```

Comment:
I wonder if it would be easy to extract an interface from `RxBleInternalScanResult` and then leverage `ScanFilter.matches(RxBleInternalScanResult scanResult)` function

Reply:
Yes, that would have been simpler...I'll look in to this